controllers/loadtest_controller.go

```diff
@@ -175,131 +176,221 @@ func (r *LoadTestReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+				log.Error(err, "could not create new pod", "pod", pod)
+				return &ctrl.Result{Requeue: true}, err
 			}
 
-			return *result, err
+			return nil, nil
 		}
-	}
-	if missingPods.Driver != nil {
-		logWithDriver := log.WithValues("driver", missingPods.Driver)
 
-		pod, err := builder.PodForDriver(missingPods.Driver)
-		if err != nil {
-			logWithDriver.Error(err, "failed to construct a pod struct for supplied driver struct")
-
-			test.Status.State = grpcv1.Errored
-			test.Status.Reason = grpcv1.ConfigurationError
-			test.Status.Message = fmt.Sprintf("failed to construct a pod for driver: %v", err)
+		for i := range missingPods.Servers {
+			logWithServer := log.WithValues("server", missingPods.Servers[i])
+
+			pod, err := builder.PodForServer(&missingPods.Servers[i])
+			if err != nil {
+				logWithServer.Error(err, "failed to construct a pod struct for supplied server struct")
+				test.Status.State = grpcv1.Errored
+				test.Status.Reason = grpcv1.ConfigurationError
+				test.Status.Message = fmt.Sprintf("failed to construct a pod for server at index %d: %v", i, err)
+				if updateErr := r.Status().Update(ctx, test); updateErr != nil {
+					logWithServer.Error(updateErr, "failed to update status after failure to construct a pod for server")
+				}
+				return ctrl.Result{}, err
+			}
 
-			if updateErr := r.Status().Update(ctx, test); updateErr != nil {
-				logWithDriver.Error(updateErr, "failed to update status after failure to construct a pod for driver")
+			if missingPods.Servers[i].Pool == nil {
+				pod.Labels[config.PoolLabel] = defaultServerPool
+			} else {
+				pod.Labels[config.PoolLabel] = *missingPods.Servers[i].Pool
 			}
 
-			return ctrl.Result{}, err
+			result, err := createPod(pod)
+			if result != nil && !kerrors.IsAlreadyExists(err) {
+				logWithServer.Error(err, "failed to create pod for server")
+				test.Status.State = grpcv1.Errored
+				test.Status.Reason = grpcv1.KubernetesError
+				test.Status.Message = fmt.Sprintf("failed to create pod for server at index %d: %v", i, err)
+				if updateErr := r.Status().Update(ctx, test); updateErr != nil {
+					logWithServer.Error(updateErr, "failed to update status after failure to create pod for server")
+				}
+				return *result, err
+			}
 		}
+		for i := range missingPods.Clients {
+			logWithClient := log.WithValues("client", missingPods.Clients[i])
+
+			pod, err := builder.PodForClient(&missingPods.Clients[i])
+			if err != nil {
+				logWithClient.Error(err, "failed to construct a pod struct for supplied client struct")
+				test.Status.State = grpcv1.Errored
+				test.Status.Reason = grpcv1.ConfigurationError
+				test.Status.Message = fmt.Sprintf("failed to construct a pod for client at index %d: %v", i, err)
+				if updateErr := r.Status().Update(ctx, test); updateErr != nil {
+					logWithClient.Error(updateErr, "failed to update status after failure to construct a pod for client")
+				}
+				return ctrl.Result{}, err
```

Comment:
There are many cases you returned an error and ctrl.Result{}, it is a bit confusing if you want to requeue the items or not, we might want to have a rule for what we return such as:
1. want to requeue item: ctrl.Result{Requeue: true}  or error, depend on what you want you could chose either one.
2. do not want to requeue item: ctrl.Result{} + nil
3. want to requeue with a specific time: ctrl.Result{RequeueAfter: Duration} + nil

Reply:
The above is my naive thought, we could come up with other rules.

Reply:
For #3, I have not checked controller-runtime again... but I'm guessing the `ctrl.Result{RequeueAfter: ...}` is ignored when the error is also returned?

Reply:
Yes, if there is a error returned with ctrl.Result{RequeueAfter: ...} then it is requeued same as  ctrl.Result{Requeue: true}

Reply:
That is crazy. Why can't you have an error with a delay to retry?

Reply:
@wanlin31 I have addressed the consistency problem in 74fda57. Let me know what you think.